### ISSUE TITLE
cell-var-from-loop (W0640): why 

### ISSUE THREAD

**Issue by bersbersbers** [PyCQA/pylint#5263]
### Bug description

I don't see why the code in this example file

https://github.com/PyCQA/pylint/blob/96e84595194073ea54a8c7730b86125049c0f4f9/tests/functional/n/nested_func_defined_in_loop.py#L1-L10

is problematic. When called using `example(range(4))`, it outputs
```
0
1
2
3
0
1
2
3
```
which looks correct to me. See also https://github.com/vald-phoenix/pylint-errors/issues/7

### Command used

```shell
wget https://raw.githubusercontent.com/PyCQA/pylint/main/tests/functional/n/nested_func_defined_in_loop.py
pylint nested_func_defined_in_loop.py
```

### Pylint output

```shell
************* Module nested_func_defined_in_loop
nested_func_defined_in_loop.py:7:18: W0640: Cell variable i defined in loop (cell-var-from-loop)

-----------------------------------
Your code has been rated at 8.57/10
```

### Expected behavior

No error

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.4
Python 3.9.6 (default, Jan  1 1970, 00:00:00) 
[GCC 7.5.0]
```

### OS / Environment

OpenSUSE Leap 15.2

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas This actually seems to be a false positive.

We have:
https://github.com/PyCQA/pylint/blob/8e15c1dd9d0dc07cb9794353a32cd37d57f2d569/tests/functional/c/cellvar_escaping_loop.py#L134-L142

But this is different because `nested` isn't called which makes the loop variable escape. That doesn't happen in the example provided here.

Looking at some of the `good_case`'s in those tests I don't see something that resembles this example. Should we allow this pattern?

**Comment by Pierre-Sassoulas** [comment]
Thank you for taking the time to look into it, I agree with you, let's add a new "good" test case.

**Comment by Pierre-Sassoulas** [comment]
#7100 should probably be done at the same time.

**Comment by ericvergnaud** [comment]
Here is another example of a false positive:

```
def some_function(samples: list[dict], *names: str):
    for name in names:
        # pylint warning W0640 here is a pylint bug (verified manually)
        filtered_samples = [s["some_key"] for s in filter(lambda s: s["some_key"]["some_name"] == name, samples)]
    # do something with filtered_samples
```
